AI #1569: Adding File Naming and Headers Conventions (#1822)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: guidelines/contributors/markdownpp-guidelines.md

Reviewer: Can we globally switch all bullets from `-` to `*`, to match our spec convention?

Author: Addressed. Resolving this thread

This feature automatically:

* Excludes marked headers from TOC generation
* Prevents section numbering for these headers
* Skips anchor link creation
* Ensures proper spacing around headers to meet markdown linting requirements
* Maintains document readability and structure

**Automatic Anchor Generation:**

All headers appearing after the !TOC directive automatically receive an HTML anchor generated from the header text whereby the following applies:

* Anchor text is normalized to lowercase.
* Whitespace and the following characters are removed: `,` `-` `(` `)`.
* Other special characters are preserved.
* Dots (.) are inserted between header levels to represent the full hierarchy of document sections.

> * Headers marked with `<!--SkipTOC-->` will appear in the document but will not be included in the Table of Contents, and no anchor will be created.

Author: ```suggestion
* Headers marked with `<!--SkipTOC-->` will appear in the document but will not be included in the Table of Contents, and no anchor will be created.

***Note:*** *These anchors are generated for the HTML and PDF builds of the specification and differ from how GitHub natively generates anchors for markdown files. Links using these anchors will work in the built specification output but not when viewing the source markdown on GitHub.*
```

This feature automatically:

* Excludes marked headers from TOC generation
* Prevents section numbering for these headers
* Skips anchor link creation
* Ensures proper spacing around headers to meet markdown linting requirements
* Maintains document readability and structure

> **Automatic Anchor Generation:**

Reviewer: Why is this not its own subheader?

```suggestion
#### Automatic Anchor Generation
```

Reviewer: Discussed.  Leaving as-is to follow the existing pattern.  Resolving.

!INCLUDE "attributes/numeric_format.md"
```

### File Naming Conventions

To ensure consistency, ease of maintenance, and portability, all Markdown and MarkdownPP file names in the FOCUS Specification repository SHOULD follow these conventions:

> - File names MUST use lowercase characters.

Reviewer: Do we want/need a carveout for ALL UPPER CASE file names?  I'm vaguely aware of a GitHub convention that ALL CAPS means it's an important file.  But for example, in our root folder, we have CHANGELOG, CONTRIBUTING, OUTPUTS, README...

Reviewer 2: Key files are all uppercase, as you've mentioned... I do support this for uppercase.

Author: I was focused on column, attribute, and dataset spec files, and totally forgot about key repository files. 😄 

changed this:

> To ensure consistency, ease of maintenance, and portability, all Markdown and MarkdownPP file names in the FOCUS Specification repository SHOULD follow these conventions:
> 
> - File names MUST use lowercase characters.

to this:

> To ensure consistency, ease of maintenance, and portability, all Markdown and MarkdownPP files in the FOCUS Specification repository, except key repository files, SHOULD follow these naming conventions:
> 
> - File names MUST use lowercase characters.
> - ...
> 
> Key repository files (e.g., `README.md`, `CHANGELOG.md`, `CONTRIBUTING.md`) SHOULD follow common GitHub naming conventions.

Author: Both the `.github` directory and its subdirectories and key repository files are excluded from the main naming convention requirements and addressed separately.

MUST is used throughout both sections, including in the introductory sentences and in the separate requirements for key repository files and the `.github` directory and its subdirectories.

Author: @ljadvey @shawnalpay can I close this thread?

Reviewer 2: Yes

- Spaces or camelCase MUST NOT be used.
- File names SHOULD remain stable after publication to maintain repository consistency and predictability.
  - If a file must be renamed, all affected `!INCLUDE` statements and build dependencies MUST be updated accordingly to avoid broken references.
- File names SHOULD be descriptive and reflect the primary content of the file.

> - Words in file names SHOULD be separated using underscores (`_`) or joined without any separator.

Reviewer 2: IMO - must with _

Please.

Reviewer 3: nit: You can merge this and the spaces requirements by just saying file names must be snake-cased and use an example (e.g., `snake_cased_file_name.md`).

Reviewer: @ljadvey 
I just want to make sure that you are in support of changing the filenames for literally every column we have.  For example, `pricingunit.md` becomes `pricing_unit.md`.  Any external links that have ever been created against columns in the repo will fail.

I'm not saying the cons outweigh the pros.  I just want to make sure that you understand that, and that you think we should do it anyway.

@ijurica phrased it the way she did in order to reflect reality.  That reality is not ideal, but it _is_ the status quo, and any change will be painful.  This is why we will need Maintainer consensus to move forward.

Reviewer 2: @shawnalpay if we are going to change things, then my vote is for _ (however, I am not firm on this and could be convinced of -).

But changing the names is a significant change.  I have worked with the way things are for a while now and can live with it.

Reviewer: @ljadvey 
Yeah, that's really the question to sort here: do we stick with the (objectively bad) status quo, or do we fix it and move everyone's cheese?

I don't think there's a right answer, but whatever the decision, we'll have to do it as a group.

Reviewer 2: @shawnalpay agreed.... worthy of discussing on the maintainer's call after async review

Author: Discussed in Maintainers meeting on Jan 19. The File Naming Conventions section has been updated according to the agreement from that meeting. I also added a Directory Naming Conventions section.

Resolving this thread.

> - Words in file names MAY be separated using hyphens (`-`).

Reviewer: We need to align on one and only one of these (either no space, `_`, or `-`).  We use a heavy mixture of all three, and it's confusing for all parties.

Let's bring it up with the Maintainers.

Reviewer 3: +1
My vote: I prefer kebab-casing (with dashes) for typing efficiency 😜

Author: Current suggestion mandates kebab-casing:

> * [Directory|File] names consisting of multiple words MUST use hyphens (`-`) as the word separator (e.g., kebab-case), except for...

I'm resolving this thread.

**Preferred file name examples:**

```text
numeric_format.md
supported_features.mdpp
cost_and_usage.md
pricingquantity.md
```

**Forbidden file name examples:**

```text

> PricingQuantity.md

Reviewer: Why is this one not allowed?  Doesn't seem to violate any of the above rules.  You specified camelCase is not allowed, but not ProperCase.

Author: Covered by the first req. Good point that I only mentioned camelCase. 

Omitting the mention of camelCase because the lowercase requirement already covers all those case variations and addressing key repository files...

Changing to:

> To ensure consistency, ease of maintenance, and portability, all Markdown and MarkdownPP files in the FOCUS Specification repository, except key repository files, SHOULD follow these naming conventions:
> 
> - File names MUST use lowercase characters.
> - Spaces MUST NOT be used.
> ...
> 
> Key repository files (e.g., `README.md`, `CHANGELOG.md`, `CONTRIBUTING.md`) SHOULD follow common GitHub naming conventions.

Reviewer: Great!  In that case, we can resolve this.

!INCLUDE "attributes/numeric_format.md"
```

### File Naming Conventions

To ensure consistency, ease of maintenance, and portability, all Markdown and MarkdownPP files in the FOCUS Specification repository, except key repository files, SHOULD follow these naming conventions:

- File names MUST use lowercase characters.
- Spaces MUST NOT be used.
- File names SHOULD remain stable after publication to maintain repository consistency and external hyperlink integrity.
  - If a file requires a rename:

>     - all affected `!INCLUDE` statements and build dependencies MUST be updated accordingly to avoid broken references.  

Reviewer 4: Do we need to link to a definition/explanation of `!INCLUDE` here? I believe it is a programming function that isn't always present in Markdown projects.

Author: The MarkdownPP docs are already linked in the Overview. 

> The FOCUS Specification uses [MarkdownPP](https://github.com/amyreese/markdown-pp) (Markdown Preprocessor)...

`!INCLUDE` is introduced shortly after and referenced in several other places, so I’d prefer not to repeat the link here.

Reviewer: Agree.  [Near the beginning of the document](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/614bc1c55d315c68ae5a6ea725ffb5caeea66127/guidelines/contributors/markdownpp-guidelines.md#markdownpp-usage-in-focus), we have a whole section explaining INCLUDEs.

No action here; already addressed in previous version IMO.

Author: @qquigley can i resolve this thread?

Author: Discussed offline with @qquigley , and he agreed to close this thread.

    - The FinOps Foundation staff MUST be notified in order to repair any hyperlink references to that file within Foundation assets.
- File names SHOULD be descriptive and reflect the primary content of the file.
- Words in file names SHOULD be separated using underscores (`_`) or joined without any separator.
- Words in file names MAY be separated using hyphens (`-`).

> Key repository files (e.g., `README.md`, `CHANGELOG.md`, `CONTRIBUTING.md`) SHOULD follow common GitHub naming conventions.

Reviewer 3: nit: I think we have an Exclusions section for attributes. Do we want to follow that pattern here? If not, should we make this a bulleted requirement?

Author: I intentionally avoid a separate Exceptions section 😄. 

The first (top-level) parent in both the Requirements sections within FOCUS entities is not a bullet, so I followed that pattern; it also aligns better with the overall style of this document.

Note: If we add a bullet point here, we should also add it to the first line in the section (“To ensure consistency, ease of maintenance, and portability, all Markdown and MarkdownPP … except …”), since they are at the same level.

Author: @flanakin, can I close this thread?

Reviewer: > I intentionally avoid a separate Exceptions section 😄.

I'm not following: I see that you _have_ included a separate Exceptions section in the latest version.

> #### Exception: Key repository files
> 
> - Key repository files (e.g., `README.md`, `CHANGELOG.md`, `CONTRIBUTING.md`) MUST follow common GitHub file naming conventions.

In any case, I think the concern has been addressed, and that we can resolve this.

> Key repository files (e.g., `README.md`, `CHANGELOG.md`, `CONTRIBUTING.md`) SHOULD follow common GitHub naming conventions.

Reviewer 3: I sould say "MUST" since these are industry best practices.
```suggestion
Key repository files (e.g., `README.md`, `CHANGELOG.md`, `CONTRIBUTING.md`) MUST follow common GitHub naming conventions.
```

Reviewer: Agree -- but what _are_ the "common GitHub naming conventions"?  Is there a link we could add?

Author: I couldn’t find a good reference for general GitHub naming conventions :( 
Given that, is using SHOULD more appropriate here?

Reviewer: @ijurica 
Here are a couple of relevant links:
https://github.com/orgs/community/discussions/22673
https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/creating-a-default-community-health-file

Author: Changed the BCP14 keyword from SHOULD to MUST.

Regarding the links - I meant to say that I couldn't find a git file naming convention reference, and I'd rather not include links with examples of key repository files or `.github` subdirectories here.

Author: resolving this thread since the original suggestion was applied

**Preferred file name examples:**

```text
numeric_format.md
supported_features.mdpp
cost_and_usage.md

> pricingquantity.md

Reviewer 3: Do we need to mention that column file names must follow the ID name and not the display name? When I look at this, it looks odd that we don't have `pricing_quantity.md` for consistency. But I do prefer the ID for simpler references.

Reviewer: @ijurica laid out examples that reflect our current reality, which is a mix of spaces, underscores, and dashes.

In the Maintainers meeting next week, I do hope that we can come to a consensus on whether to change that to a single format.  That would then inform examples like this shifting.

Author: Updated version reflects the agreement reached in the Maintainers meeting on Jan 19. Resolving the thread.

**Preferred directory name examples:**

```text
supported-features
costandusage
pricingquantity
numericformat
```

**Forbidden directory name examples:**

```text
SupportedFeatures
Supported Features
supportedFeatures
supported_features
cost-and-usage
```

### File Naming Conventions

To ensure consistency, ease of maintenance, and portability, all Markdown and MarkdownPP files in the FOCUS Specification repository, except key repository files, MUST follow these naming conventions:

- File names MUST use lowercase characters.
- File names MUST NOT contain spaces.
- File names SHOULD remain stable after publication to maintain repository consistency and external hyperlink integrity.
  - If a file requires a rename:
    - all affected `!INCLUDE` statements and build dependencies MUST be updated accordingly to avoid broken references.
    - The FinOps Foundation staff MUST be notified in order to repair any hyperlink references to that file within Foundation assets.
- File names SHOULD be descriptive and reflect the primary content of the file.

> - File names consisting of multiple words MUST use hyphens (`-`) as the word separator (e.g., kebab-case.md), except for files defining FOCUS entities (e.g., FOCUS datasets, FOCUS columns, FOCUS attributes).

Author: ```suggestion
- File names consisting of multiple words MUST use hyphens (`-`) as the word separator (e.g., kebab-case), except for files defining FOCUS entities (e.g., FOCUS datasets, FOCUS columns, FOCUS attributes).
```

!INCLUDE "attributes/numeric_format.md"
```

### Directory Naming Conventions

To ensure consistency, ease of maintenance, and portability, all directories in the FOCUS Specification repository, except the `.github` directory and its subdirectories, MUST follow these naming conventions:

- Directory names MUST use lowercase characters.
- Directory names MUST NOT contain spaces.
- Directory names SHOULD be descriptive and reflect the primary content of the directory.

> - Directory names consisting of multiple words MUST use hyphens (`-`) as the word separator (e.g., kebab-case), except for directories defining FOCUS entities (e.g., FOCUS datasets, FOCUS columns, FOCUS attributes).

Reviewer: ```suggestion
- Directory names consisting of multiple words MUST adhere to the following requirements:
  - Directory names for entities with identifiers defined in the FOCUS specification (e.g., datasets, columns, attributes) MUST NOT separate words (e.g., `contractcommitment`).
  - Directory names for all other entities MUST separate words with hyphens (`-`) (e.g., `requirements-model`).
```

Author: I hope I've managed to addressed this concern in the latest version. 
Please let me know if it’s okay to close this thread.

Reviewer: Looks good to me.  Resolving.

- Directory names consisting of multiple words MUST NOT use any word separator for directories defining FOCUS entities, so that the directory name matches the FOCUS entity identifier in lowercase.

#### Exception: GitHub directories

> - The `.github` directory and its subdirectories (e.g., `ISSUE_TEMPLATE`, `workflows`) MUST follow common GitHub directory naming conventions.

Reviewer 5: The inline "except..." in the descriptor and bullets duplicates what's already in the Exception section. Cleaner to keep the rules simple and let the exceptions section handle the carve-outs — matches how we do it in the spec for attributes.

```suggestion
To ensure consistency, ease of maintenance, and portability, all directories in the FOCUS Specification repository MUST follow these naming conventions. See Directory naming exceptions below for cases where these conventions do not apply.

- Directory names MUST use lowercase characters.
- Directory names MUST NOT contain spaces.
- Directory names SHOULD be descriptive and reflect the primary content of the directory.
- Directory names consisting of multiple words MUST use hyphens (`-`) as the word separator (e.g., kebab-case).

#### Directory naming exceptions

- The `.github` directory and its subdirectories (e.g., `ISSUE_TEMPLATE`, `workflows`) MUST follow common GitHub directory naming conventions.
- Directories defining FOCUS entities (e.g., FOCUS datasets, FOCUS columns, FOCUS attributes) MUST NOT use any word separator, so that the directory name matches the FOCUS entity identifier in lowercase.
```

Reviewer: @thecloudman Looks like you're calling out something similar I was trying to address with [this thread](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1822/files#r2962666769) (looks like we posted around the same time yesterday).  

@ijurica In any case, looks like Graham and I are in agreement that something needs tweaking here.

Author: @thecloudman @shawnalpay: thanks both! I understand the concern.

I initially tried to combine both of your suggestions, but I ultimately decided to **separate FOCUS-defined rules from externally governed cases** (e.g., GitHub conventions), rather than using a general “exceptions” section, after all (see the latest version)

The main reason is to **avoid ambiguity around which rules the exceptions apply to** (e.g., whether they override all rules or only specific ones like word separators; for instance, we enforce lowercase for `costandusage`, but not for GitHub-managed directories such as `ISSUE_TEMPLATE` or `workflows`). This way, the scope of each rule set is explicit, and we avoid implicit or partial overrides — which is also why we are **removing the Exceptions section in Attributes**.

I’ve applied this consistently across both directory and file naming. 

I hope you’ll find this an acceptable balance between clarity and separation of governance. Please let me know if I can resolve this thread or if further adjustments are required.

Reviewer: I'm good with it.  I resolved my thread on this topic.  Will leave it to @thecloudman to resolve this one.

Reviewer 5: @ijurica I am happy with this proposal.

#### Preferred directory name examples

```text
supported-features
costandusage
pricingquantity
numericformat
```

#### Forbidden directory name examples

```text
SupportedFeatures
Supported Features
supportedFeatures
supported_features
cost-and-usage
```

### File Naming Conventions

To ensure consistency, ease of maintenance, and portability, all Markdown and MarkdownPP files in the FOCUS Specification repository, except key repository files, MUST follow these naming conventions:

- File names MUST use lowercase characters.
- File names MUST NOT contain spaces.
- File names SHOULD be descriptive and reflect the primary content of the file.
- File names consisting of multiple words MUST use hyphens (`-`) as the word separator (e.g., kebab-case), except for files defining FOCUS entities (e.g., FOCUS datasets, FOCUS columns, FOCUS attributes).
- File names consisting of multiple words MUST NOT use any word separator for files defining FOCUS entities, so that the file name matches the FOCUS entity identifier in lowercase.

#### Exception: Key repository files

> - Key repository files (e.g., `README.md`, `CHANGELOG.md`, `CONTRIBUTING.md`) MUST follow common GitHub file naming conventions.

Reviewer 5: Same thing here — move the exception language out of the descriptor and bullets into the exceptions section.

```suggestion
To ensure consistency, ease of maintenance, and portability, all Markdown and MarkdownPP files in the FOCUS Specification repository MUST follow these naming conventions. See File naming exceptions below for cases where these conventions do not apply.

- File names MUST use lowercase characters.
- File names MUST NOT contain spaces.
- File names SHOULD be descriptive and reflect the primary content of the file.
- File names consisting of multiple words MUST use hyphens (`-`) as the word separator (e.g., kebab-case).

#### File naming exceptions

- Key repository files (e.g., `README.md`, `CHANGELOG.md`, `CONTRIBUTING.md`) MUST follow common GitHub file naming conventions.
- Files defining FOCUS entities (e.g., FOCUS datasets, FOCUS columns, FOCUS attributes) MUST NOT use any word separator, so that the file name matches the FOCUS entity identifier in lowercase.
```

Author: An updated version is available. I’ve tried to address your concern but ultimately decided to separate FOCUS-defined rules from externally governed cases (see [this thread](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1822/files/c1be9d9b08630708d0b12f6a58f907a91b66338c#r2969767655) for details). 

Please let me know if I can resolve this thread or if further adjustments are required.

Reviewer 5: @ijurica  i have read the updated thread and am okay with the updates, i will resolve this

> !INCLUDE "attributes/numeric_format.md"

Reviewer: Just to confirm... we're going to apply these changes in a separate PR?  Gemini is getting mad that we're not following the guidelines, even within this PR.  

I believe our plan is to do this during consistency review.  If so, you can resolve.

### 🚨 Critical Contradiction: Existing Examples vs. New Rules
The newly added rules explicitly forbid using underscores in filenames and directories. In fact, `numeric_format.md` is specifically listed under **Forbidden File Name Examples**. 

However, right above where your new additions start (around line 193 of the existing file), there is an existing code block demonstrating an include statement:
`!INCLUDE "attributes/numeric_format.md"`

**The Fix:** You need to update that existing code block in your diff so it reflects the new standard (e.g., changing it to `!INCLUDE "attributes/numericformat.md"`), otherwise the document contradicts itself immediately before laying out the rules.

Author: My assumption is that the repository structure snapshot and examples of !INCLUDE statements in this document (like !INCLUDE "attributes/numeric_format.md") should be updated after we apply the naming conventions to the actual repository — not before.

Considering we have AI #1868, is it ok if I resolve this?

---

BTW, just added [this reminder](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/issues/1868#issuecomment-4106201635) in AI #1868 - hope that's ok.

Reviewer: Sounds good.

```

### Directory Naming Conventions

> To ensure consistency, ease of maintenance, and portability, all directories in the FOCUS Specification repository, except [GitHub-managed directories](#github-managed-directories), MUST follow these naming conventions:

Reviewer: ### Self-Contradictory Internal Links (Broken Anchors)
In your new text, you create two internal links to sections within the same document:
* `[GitHub-managed directories](#github-managed-directories)`
* `[key repository files](#key-repository-files)`

However, according to your newly defined **Automatic Anchor Generation** rules:
> * Whitespace and the following characters are removed: `,` `-` `(` `)`.

If whitespace and hyphens are truly *removed* by your build process, the generated anchors for those headers would actually be `#githubmanageddirectories` and `#keyrepositoryfiles`. Furthermore, your rules state that dots are inserted between header levels, meaning the anchors might be even more complex depending on the parent sections.

**The Fix:** You need to either update your internal links in this document so they match the exact output of your custom anchor generation, or revise the anchor generation rules to reflect standard GitHub Markdown behavior (where spaces become hyphens and existing hyphens are preserved).

Author: The links aren't actually broken. The guidelines documents live outside specification/ and are rendered by GitHub, so standard GitHub anchor generation applies — which works correctly for these links.

The automatic anchor generation section describes how the modified `TableOfContents.py` creates anchor links for the HTML and PDF builds of the specification — it's **not a set of rules to follow**, but a description of what the script does so contributors know what to expect without reading the Python code. These anchors differ from standard GitHub anchor generation and only apply to content within specification/.

But I agree it was confusing, so I've now added an explicit note about this in the document:

> ***Note:*** *These anchors are generated for the HTML and PDF builds of the specification and differ from how GitHub natively generates anchors for markdown files. Links using these anchors will work in the built specification output but not when viewing the source markdown on GitHub.*

@shawnalpay, please let me know if the added note makes it clear enough what these anchors apply to and how they differ from standard GitHub behavior.

---

BTW, naming conventions have been extracted into a standalone document (`repository-naming-conventions.md`), and `markdownpp-guidelines.md` has an explicit scope note clarifying it applies to the `specification/` directory. (See the [PR description](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1822#issue-3805962485) for more details and reasoning)

Reviewer: Great, thanks!  Resolving.

supported-features
costandusage
pricingquantity
numericformat
```

#### Forbidden directory name examples

```text
SupportedFeatures
Supported Features
supportedFeatures
supported_features
cost-and-usage
```

### File Naming Conventions

To ensure consistency, ease of maintenance, and portability, all Markdown and MarkdownPP files in the FOCUS Specification repository, except key repository files, MUST follow these naming conventions:

- File names MUST use lowercase characters.
- File names MUST NOT contain spaces.
- File names SHOULD be descriptive and reflect the primary content of the file.
- File names consisting of multiple words MUST use hyphens (`-`) as the word separator (e.g., kebab-case), except for files defining FOCUS entities (e.g., FOCUS datasets, FOCUS columns, FOCUS attributes).
- File names consisting of multiple words MUST NOT use any word separator for files defining FOCUS entities, so that the file name matches the FOCUS entity identifier in lowercase.

#### Exception: Key repository files

- Key repository files (e.g., `README.md`, `CHANGELOG.md`, `CONTRIBUTING.md`) MUST follow common GitHub file naming conventions.

#### Renaming files after publication

> File names are used by build processes (e.g., `!INCLUDE` statements) and may also appear in external hyperlinks (e.g., within FinOps Foundation assets). File names SHOULD remain stable after publication to maintain repository consistency and external hyperlink integrity.

Author: We’ve added notes about notifying the FinOps Foundation staff in case of renaming (directories, files, headers), assuming these may be referenced in Foundation assets. 

@Matt-Cowsert, @mike-finopsorg, @shawnalpay - two questions:

1. Can we **confirm** how the specification is referenced externally today (e.g., links to files vs. rendered HTML vs. section anchors)? Based on that, we can validate whether this requirement is needed and consistent across all three cases.

2. Do we expect contributors to notify the FinOps Foundation staff when such changes occur? If not, should we remove the following guideline (currently present in all three sections):

> - The FinOps Foundation staff MUST be notified to repair any hyperlink references within Foundation assets.

Reviewer: @ijurica 
1. We cannot confirm any external reference to the spec, be it Foundation assets or otherwise.  Anecdotally, I can tell you that all of those are in use in various artifacts.  You should keep references to all.
2. A contributor does not need to explicitly notify Foundation staff.  We are inherently embedded in the working group review and approval process, and it's incumbent upon us to review PRs and own any relevant updates; no need for a contributor to take some extra path to notifying us.

Author: Thanks!

Based on this, I've:

- removed the requirement for contributors to notify FinOps Foundation staff across all three sections.
- kept the references to external usage (e.g., Foundation assets) and the guidance to keep names stable after publication, to reflect that these links may exist.

Reviewer: Looks good!  Resolving.

<!-- Example: # Title becomes ### Title -->
```

### Header Conventions

FOCUS Specification relies on Markdown headers to define document structure, section numbering, and navigation. Consistent header usage is required to ensure predictable Table of Contents (TOC) generation and stable internal links.

#### Automatic Anchor Generation

All headers appearing after the !TOC directive automatically receive an HTML anchor generated from the header text whereby the following applies

- Anchor text is normalized to lowercase.
- Whitespace and the following characters are removed: `,` `-` `(` `)`.
- Other special characters such as `/`, `:`, `%`, `&`, `@` are preserved.
- Dots (.) are inserted between header levels to represent the full hierarchy of document sections.
- Headers marked with `<!--SkipTOC-->` MAY appear in the document but will not be included in the Table of Contents, and no anchor will be created.

> 

Reviewer: ```suggestion
- Duplicate header names are automatically de-duplicated for anchor generation.  For example: the first use of header `Column A` results in `#columna`, whereas the second use results in `#columna-1`.

```

<!-- Example: # Title becomes ### Title -->
```

### Header Conventions

> FOCUS Specification relies on Markdown headers to define document structure, section numbering, and navigation. Consistent header usage is required to ensure predictable Table of Contents (TOC) generation and stable internal links.

Reviewer 4: ```suggestion
The FOCUS Specification relies on Markdown headers to define document structure, section numbering, and navigation. Consistent header usage is required to ensure predictable Table of Contents (TOC) generation and stable internal links.
```

Reviewer 4: Nit I can walk away from.

Author: I personally prefer the original (i.e., without the article), but I’d like to get at least one more opinion before resolving this one way or the other.

Reviewer: In English informative sections like this, we should use proper grammar -- and in this case, @qquigley is correct that we need a definite article here.

Agree with the change.

Author: I've applied this suggestion manually, as it couldn't be committed directly from the suggestion box.
resolving

#### Automatic Anchor Generation

All headers appearing after the !TOC directive automatically receive an HTML anchor generated from the header text whereby the following applies:

- Anchor text is normalized to lowercase.
- Whitespace and the following characters are removed: `,` `-` `(` `)`.
- Other special characters such as `/`, `:`, `%`, `&`, `@` are preserved.
- Dots (.) are inserted between header levels to represent the full hierarchy of document sections.
- Headers marked with `<!--SkipTOC-->` will appear in the document but will not be included in the Table of Contents, and no anchor will be created.

> - Duplicate header names are automatically de-duplicated for anchor generation.  For example: the first use of header `Column A` results in `#columna`, whereas the second use results in `#columna-1`.

Reviewer 3: nit: I would say that this is non-standard. Generally, anchors are rendered using kebab-casing. If this is how it works today, that's fine. But just want to call out that this is atypical.

Reviewer: That's how it's always worked within MarkdownPP before @mike-finopsorg's recent change to add namespaces to anchors.  

<img width="1133" height="498" alt="Screenshot 2026-01-14 at 7 44 19 PM" src="https://github.com/user-attachments/assets/d8b82855-2a4e-4802-8ba4-3ab4da3e391c" />

Glancing through the spec HTML now, however, I don't see any duplicates anymore.

@mike-finopsorg, should we strike this line, or is there still dedupe logic in the TOC generator anywhere?

Author: I've removed the de-duplication bullet point from Automatic Anchor Generation.

After checking the latest version of TableOfContents.py, I've realized that de-duplication was replaced by hierarchy-based anchor generation, i.e., the two are not coexisting.

The script change and the initial draft of this PR were made within a few days of each other, so I was clearly working off the old version of the script at some point 😄

Reviewer 6: Yeah @shawnalpay markdown-pp has always handled the anchors like this so I think we should resolve as it's not a major issue how it works today.

Reviewer: Sounds good all, thanks!  Resolving with no action.

#### File-Level Header Structure

- Content-oriented Markdown files (.md) SHOULD begin with a top-level header that defines the section title.
- MarkdownPP assembly files (.mdpp) that consist only of !INCLUDE directives MAY omit headers.

#### Header Syntax

- Headers MUST use ATX-style Markdown syntax (`#`).
- Up to six header levels (######) are supported.
- A single space MUST follow the `#` character(s).
- Headers MUST NOT use Setext-style headers (`=` or `-` underlines), even though they are technically supported.
- Headers MUST NOT be empty.

> - Special characters that may interfere with anchor generation SHOULD be used with caution.

Reviewer 3: nit: I'm not sure how clear this is. I would flip it and say that the headers SHOULD only include alphanumeric text and avoid special characters other than dash, apostrophe, and parens. Not sure how many others we'd need, but we could expand the list as needed.

Reviewer: I'm in favor!

Author: ```suggestion
- Header text SHOULD only contain alphanumeric characters (`A-Z`, `0-9`), spaces (` `), commas (`,`), and the following special characters: hyphens (`-`), apostrophes (`'`), and parentheses (`(`, `)`). Other special characters SHOULD be avoided.
```

Author: @shawnalpay @flanakin please provide feedback on this suggestion or propose an alternative. Thanks!

Reviewer: Agree with your proposal @ijurica!  Ship it.

<!-- Example: # Title becomes ### Title -->
```

> ### Automatic Anchor Generation

Reviewer: ### Missing Character Rule (Apostrophes)
Under **Header Syntax**, you explicitly allow apostrophes (`'`):
> ...and the following special characters: hyphens (`-`), apostrophes (`'`), and parentheses (`(`, `)`).

However, apostrophes are completely missing from the **Automatic Anchor Generation** rules. You explain what happens to whitespace, commas, hyphens, and parentheses (they are removed), and what happens to `/`, `:`, etc. (they are preserved), but you don't mention what happens to the allowed apostrophe.

Author: I hear you. 

The apostrophe falls under the general rule and is preserved. I mentioned it as an allowed character in header syntax but missed including it as one of the examples of preserved characters in the anchor generation section. 

Decided to simplify to just "Other special characters are preserved" to avoid listing some characters but missing others. 

Hope you agree with the decision.

Reviewer: Sounds good.  Resolving!
